File: Spoj/AGGRCOW.cpp
#include <iostream>

using namespace std;

int getLargestMinDist(int A[], int N, int mid, int K)
{
    int count = 1;
    int i = 0;
    int j = 1;
    while (j < N)
    {
        int diff = A[j] - A[i];
        if (diff >= mid)
        {
            count++;
            i = j;
        }
        // cout << i << "  " << diff << "\n";
        if (count == K)
        {
            return true;
        }
        j++;
    }
    return false;
}

int bst(int A[], int N, int K)
{
    int low = 0, high = 1000000000;
    while (low <= high)
    {
        int mid = (low + high) / 2;
        bool flag = getLargestMinDist(A, N, mid, K);
        if (flag == true)
        {
            low = mid + 1;
        }
        else
        {
            high = mid - 1;
        }
        // cout << low << " " << mid << " " << high << "  " << flag << "\n";
    }
    return high;
}

void solve()
{
    int T;
    cin >> T;
    while (T--)
    {
        int N, K;
        cin >> N >> K;
        int A[N];
        for (int i = 0; i < N; i++)
        {
            int a;
            cin >> a;
            A[i] = a;
        }
        sort(A, A + N);
        cout << bst(A, N, K) << "\n";
    }
}

int main()
{

    freopen("/Users/gopaljaiswal/project/frameworks/Anaconda_python/Comp_Prog/"
            "input.txt",
            "r", stdin);
    freopen("/Users/gopaljaiswal/project/frameworks/Anaconda_python/Comp_Prog/"
            "output.txt",
            "w", stdout);

    solve();
    return 0;
}

File: Spoj/AGGRCOW.py
A = []


def isPosible(M, t):
    count = 1
    i = 0
    for j in range(1, len(A)):
        diff = A[j]-A[i]
        if diff >= t:
            count += 1
            i = j
            # if diff == t:
            #     diffVar = diff

    return count


def getRes(M):
    res = 1
    low, high = 1, 10**9

    # count, diffVar = map(int, isPosible(M, 3).split())
    # print str(count)+" "+str(diffVar)
    # print str(isPosible(M, 3))+"  "+str(diffVar)

    while(low <= high):
        mid = (low+high) >> 1
        # count, diffVar = map(int, isPosible(M, mid).split())
        count = isPosible(M, mid)
        # print 'low={}, high={}, mid={}, count={}'.format(low, high, mid, count)
        if count >= M:
            res = mid
            # print str(low)+"  "+str(mid)+" "+str(high) +"  "+str(diffVar)
            low = mid+1
        else:
            high = mid-1

    print res


T = input()
for __ in range(T):
    A = []
    N, M = map(int, raw_input().split())

    for i in range(0, N):
        A.append(input())
    A.sort()
    getRes(M)
